respy/state_space.py

```diff
@@ -0,0 +1,430 @@
+                e.__traceback__
+            )
+
+        return attribute[indices]
+
+    def _create_slices_by_periods(self, n_periods):
+        """Create slices to index all attributes in a given period.
+
+        It is important that the returned objects are not fancy indices. Fancy indexing
+        results in copies of array which decrease performance and raise memory usage.
+
+        """
+        self.slices_by_periods = []
+        for i in range(n_periods):
+            idx_start, idx_end = np.where(self.states[:, 0] == i)[0][[0, -1]]
+            self.slices_by_periods.append(slice(idx_start, idx_end + 1))
+
+
+def _create_state_space(options, n_types):
+    """Create the state space.
+
+    This process involves two steps. First, the core state space is created which
+    abstracts from different initial experiences and instead uses the minimum initial
+    experience per choice.
+
+    Secondly, the state space is adjusted by all combinations of initial experiences and
+    also filtered, excluding invalid states.
+
+    """
+    data = _create_core_state_space(options, n_types)
+
+    exp_cols = [f"exp_{sector}" for sector in options["choices_w_exp"]]
+
+    df = pd.DataFrame.from_records(
+        data, columns=["period"] + exp_cols + ["lagged_choice", "type"]
+    )
+
+    codes_to_choices = {i: sec for i, sec in enumerate(options["choices"])}
+    df.lagged_choice = df.lagged_choice.astype("category").cat.rename_categories(
+        codes_to_choices
+    )
+
+    df = _adjust_state_space_with_different_starting_experiences(df, options)
+
+    df = df.sort_values("period").reset_index(drop=True)
+
+    indexer = _create_indexer(df, n_types, options)
+
+    return df, indexer
+
+
+def _create_core_state_space(options, n_types):
+    """Create the core state space.
+
+    The core state space abstracts from initial experiences and uses the maximum range
+    between initial experiences and maximum experiences.
+
+    """
+    minimal_initial_experience = np.array(
+        [np.min(options["sectors"][sec]["start"]) for sec in options["choices_w_exp"]]
+    )
+    additional_exp = options["maximum_exp"] - minimal_initial_experience
+
+    data = []
+    for period in range(options["n_periods"]):
+        data += list(
+            _create_core_state_space_per_period(
+                period, additional_exp, n_types, options
+            )
+        )
+
+    return data
+
+
+def _create_core_state_space_per_period(
+    period, additional_exp, n_types, options, experiences=None, pos=0
+):
+    """Create core state space per period.
+
+    First, this function returns a state combined with all possible lagged choices and
+    types.
+
+    Secondly, if there exists a choice with experience in ``additional_exp[pos]``, loop
+    over all admissible experiences, update the state and pass it to the same function,
+    but moving to the next choice which accumulates experience.
```

Comment:
You already document this function quite nicely, but maybe we should add a section to the docstring that explains

- How the loops in the old state space looked like and why we can't use such a solution anymore 
- why itertools.product, np.meshgrid,  or other more obvious solutions could not be used